transaction/memo.go

```diff
@@ -0,0 +1,144 @@
+package transaction
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/gagliardetto/solana-go"
+	solanarpc "github.com/gagliardetto/solana-go/rpc"
+)
+
+const BxMemoMarkerMsg = "Powered by bloXroute Trader Api"
+
+var TraderAPIMemoProgram = solana.MustPublicKeyFromBase58("HQ2UUt18uJqKaQFJhgV9zaTdQxUZjNrsKFgoEDquBkcx")
+
+// CreateTraderAPIMemoInstruction generates a transaction instruction that places a memo in the transaction log
+func CreateTraderAPIMemoInstruction(msg string) solana.Instruction {
+	if msg == "" {
+		msg = BxMemoMarkerMsg
+	}
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(msg))
+
+	instruction := &solana.GenericInstruction{
+		AccountValues: nil,
+		ProgID:        TraderAPIMemoProgram,
+		DataBytes:     buf.Bytes(),
+	}
+
+	return instruction
+}
+
+func AddMemo(instructions []solana.Instruction, memoContent string, blockhash solana.Hash, owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+
+	instructions = append(instructions, memo)
+
+	txnBytes, err := buildPartialSignedTxn(blockhash, owner, instructions, privateKeys)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(txnBytes), nil
+}
+
+func AddMemoToSerializedTxn(programIDIndex uint16, unsignedTxBase64, memoContent string,
+	owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	unsignedTxBytes, err := solanarpc.DataBytesOrJSONFromBase64(unsignedTxBase64)
+	if err != nil {
+		return "", err
+	}
+	unsignedTx := solanarpc.TransactionWithMeta{Transaction: unsignedTxBytes}
+	solanaTx, err := unsignedTx.GetTransaction()
+	if err != nil {
+		return "", err
+	}
+	progID, err := solanaTx.Message.ResolveProgramIDIndex(programIDIndex)
+	if err != nil {
+		return "", err
+	}
+
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+	var instructions []solana.Instruction
+
+	for _, cmpInst := range solanaTx.Message.Instructions {
+		accounts := cmpInst.ResolveInstructionAccounts(&solanaTx.Message)
+		//decodedData, err := base58.Decode(cmpInst.Data)
+		instructions = append(instructions, &solana.GenericInstruction{
+			AccountValues: accounts,
+			ProgID:        progID,
+			DataBytes:     cmpInst.Data,
+		})
+	}
+	instructions = append(instructions, memo)
+	txnBytes, err := buildPartialSignedTxn(solanaTx.Message.RecentBlockhash, owner, instructions, privateKeys)
```

Comment:
think there's no reason for partial signing here –– user should provide all the private keys that they used for the initial transaction and can fully complete signing